velox/exec/Unnest.cpp

```diff
@@ -111,56 +112,111 @@ RowVectorPtr Unnest::getOutput() {
   }
 
   const auto size = input_->size();
-  const auto maxOutputSize = outputBatchRows();
 
   // Limit the number of input rows to keep output batch size within
-  // 'maxOutputSize' if possible. Process each input row fully. Do not break
-  // single row's output into multiple batches.
-  vector_size_t numInput = 0;
+  // 'maxOutputSize' if possible. Not process each input row fully when a
+  // row's output exceeds 'maxOutputSize'. The output of single row may be
+  // divided into multiple batches.
   vector_size_t numElements = 0;
-  for (auto row = nextInputRow_; row < size; ++row) {
-    numElements += rawMaxSizes_[row];
-    ++numInput;
-
-    if (numElements >= maxOutputSize) {
-      break;
-    }
-  }
-
+  vector_size_t partialProcessRowStart = -1;
+  auto rowRange = extractRowRange(size, numElements, partialProcessRowStart);
   if (numElements == 0) {
     // All arrays/maps are null or empty.
     input_ = nullptr;
     nextInputRow_ = 0;
     return nullptr;
   }
 
-  auto output = generateOutput(nextInputRow_, numInput, numElements);
-
-  nextInputRow_ += numInput;
+  auto output = generateOutput(rowRange, numElements);
+  if (partialProcessRowStart != -1) {
+    firstRowStart_ = partialProcessRowStart;
+    nextInputRow_ += rowRange.size - 1;
+  } else {
+    firstRowStart_ = 0;
+    nextInputRow_ += rowRange.size;
+  }
 
-  if (nextInputRow_ >= size) {
+  if (nextInputRow_ >= input_->size()) {
     input_ = nullptr;
     nextInputRow_ = 0;
   }
 
   return output;
 }
 
-void Unnest::generateRepeatedColumns(
-    vector_size_t start,
+const Unnest::RowRange Unnest::extractRowRange(
     vector_size_t size,
+    vector_size_t& numElements,
+    vector_size_t& partialProcessRowStart) {
+  VELOX_DCHECK_LT(nextInputRow_, size);
+
+  vector_size_t numInput = 0;
+  vector_size_t firstRowEnd = rawMaxSizes_[nextInputRow_];
+  vector_size_t lastRowEnd = -1;
+  // May split the first row and the last row, not split middle rows.
+  // The end row related variables firstRowEnd and lastRowEnd will not be used
+  // if there is just one row, its start is always 0.
+  for (auto row = nextInputRow_; row < size; ++row) {
+    bool isFirstRow = (row == nextInputRow_);
+    vector_size_t remainingSize =
+        isFirstRow ? firstRowEnd - firstRowStart_ : rawMaxSizes_[row];
+    if (numElements + remainingSize > maxOutputSize_) {
+      // Single row's output is divided into multiple batches.
+      // Computes the row range to process the first and last row
+      // partially instead of 0 to rawMaxSizes_[row].
+      if (isFirstRow) {
+        firstRowEnd = firstRowStart_ + maxOutputSize_ - numElements;
+        partialProcessRowStart = firstRowEnd;
+      } else {
+        lastRowEnd = maxOutputSize_ - numElements;
+        partialProcessRowStart = lastRowEnd;
```

Comment:
How do we know if this row is the last row so as to avoid splitting middle rows? Thanks.

Reply:
If we process the row, the output batch size is up to maxBatchSize after we process this rows's output, it is last row.
For a row size,
sizeIdx = 0 is first row, sizeIdx = size - 1 is last row, other row is middle rows.